ToolBox/Polygon/PolygonDataQueueHandler.cs

```diff
@@ -0,0 +1,536 @@
+            {
+                Log.Error($"PolygonDataQueueHandler.ProcessHistoryRequests(): unsupported resolution: {request.Resolution}.");
+                yield break;
+            }
+
+            // check security type
+            if (request.Symbol.SecurityType != SecurityType.Equity &&
+                request.Symbol.SecurityType != SecurityType.Forex &&
+                request.Symbol.SecurityType != SecurityType.Crypto)
+            {
+                Log.Error($"PolygonDataQueueHandler.ProcessHistoryRequests(): unsupported security type: {request.Symbol.SecurityType}.");
+                yield break;
+            }
+
+            // check tick type
+            if (request.TickType != TickType.Trade)
+            {
+                Log.Error("PolygonDataQueueHandler.ProcessHistoryRequests(): Only history requests for trade bars are supported.");
+                yield break;
+            }
+
+            var ticker = request.Symbol.ID.Symbol;
+            var start = request.StartTimeUtc.ConvertFromUtc(TimeZones.NewYork);
+            var end = request.EndTimeUtc.ConvertFromUtc(TimeZones.NewYork);
+
+            Log.Trace("PolygonDataQueueHandler.ProcessHistoryRequests(): Submitting request: " +
+                      Invariant($"{request.Symbol.SecurityType}-{ticker}: {request.Resolution} {start}->{end}")
+            );
+
+            var historyTimespan = GetHistoryTimespan(request.Resolution);
+
+            var startDate = start.Date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
+            var endDate = end.Date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
+
+            var tickerPrefix = string.Empty;
+            if (request.Symbol.SecurityType == SecurityType.Forex)
+            {
+                tickerPrefix = "C:";
+            }
+            else if (request.Symbol.SecurityType == SecurityType.Crypto)
+            {
+                tickerPrefix = "X:";
+            }
+
+            // Download and parse data
+            using (var client = new WebClient())
+            {
+                var url = $"{HistoryBaseUrl}/aggs/ticker/{tickerPrefix}{ticker}/range/1/{historyTimespan}/{startDate}/{endDate}?apiKey={_apiKey}";
+
+                var response = client.DownloadString(url);
+
+                var result = JsonConvert.DeserializeObject<AggregatesResponse>(response);
+                if (result.Results == null)
+                {
+                    yield break;
+                }
+
+                foreach (var row in result.Results)
+                {
+                    var utcTime = Time.UnixMillisecondTimeStampToDateTime(row.Timestamp);
+
+                    if (utcTime < request.StartTimeUtc ||
+                        utcTime > request.EndTimeUtc.Add(request.Resolution.ToTimeSpan()))
+                    {
+                        continue;
+                    }
+
+                    var time = utcTime.ConvertFromUtc(TimeZones.NewYork);
```

Comment:
TimeZone should be exchange I think here? Thinking about Crypto/Forex for example. And the `PolygonDataDownloader` should be using Data time zone for storing the data, right?
Just checking, for forex we only have quote streamed data, but history are trades?

Reply:
I think there seems to be NBBO quotes history https://polygon.io/docs/#get_v2_ticks_stocks_nbbo__ticker___date__anchor?